dataset/src/main/scala/frameless/functions/Udf.scala

```diff
@@ -0,0 +1,75 @@
+package frameless
+package functions
+
+import org.apache.spark.sql.{Column, FramelessInternals}
+
+trait Udf {
+  def udf[T, A, R](
+     f: A => R)
+  (implicit
+     aenc: TypedEncoder[A],
+     reng: TypedEncoder[R]) =
+    (u: TypedColumn[T, A]) => {
+      val uf = FramelessInternals.makeUDF(f,reng.targetDataType,Some(Seq(aenc.targetDataType)))
+      val x: Column = uf(u.untyped)
+      new TypedColumn[T,R](x)
+    }
+
+  def udf[T, A1, A2, R](
+     f: (A1,A2) => R)
+  (implicit
+     a1enc: TypedEncoder[A1],
+     a2enc: TypedEncoder[A2],
+     reng: TypedEncoder[R]) =
+    (u1: TypedColumn[T, A1], u2: TypedColumn[T, A2]) => {
+      val uf = FramelessInternals.makeUDF(f, reng.targetDataType,Some(Seq(a1enc.targetDataType, a2enc.targetDataType)))
+      val x: Column = uf(u1.untyped, u2.untyped)
+      new TypedColumn[T,R](x)
+    }
+
+  def udf[T, A1, A2, A3, R](
+  f: (A1,A2,A3) => R)
+  (implicit
+   a1enc: TypedEncoder[A1],
+   a2enc: TypedEncoder[A2],
+   a3enc: TypedEncoder[A3],
+   reng: TypedEncoder[R]) =
+    (u1: TypedColumn[T, A1], u2: TypedColumn[T, A2], u3: TypedColumn[T, A3]) => {
+      val uf = FramelessInternals.makeUDF(f, reng.targetDataType,
+        Some(Seq(a1enc.targetDataType, a2enc.targetDataType, a3enc.targetDataType)))
+      val x: Column = uf(u1.untyped, u2.untyped, u3.untyped)
+      new TypedColumn[T,R](x)
+    }
+
+  def udf[T, A1, A2, A3, A4, R](
+   f: (A1,A2,A3,A4) => R)
+   (implicit
+    a1enc: TypedEncoder[A1],
+    a2enc: TypedEncoder[A2],
+    a3enc: TypedEncoder[A3],
+    a4enc: TypedEncoder[A4],
+    reng: TypedEncoder[R]) =
+    (u1: TypedColumn[T, A1], u2: TypedColumn[T, A2], u3: TypedColumn[T, A3], u4: TypedColumn[T, A4]) => {
+      val uf =  FramelessInternals.makeUDF(f, reng.targetDataType,
+        Some(Seq(a1enc.targetDataType, a2enc.targetDataType, a3enc.targetDataType, a4enc.targetDataType)))
+      val x: Column = uf(u1.untyped, u2.untyped, u3.untyped, u4.untyped)
+      new TypedColumn[T,R](x)
+    }
+
+  def udf[T, A1, A2, A3, A4, A5, R](
+    f: (A1,A2,A3,A4,A5) => R)
+    (implicit
```

Comment:
I was thinking about this, we don't actually need encoders for `A1` because having `u1` already implies we have encoder available as `u1.encoder`, we can do like this:
https://github.com/adelbertc/frameless/blob/master/dataset/src/main/scala/frameless/ops/GroupByOps.scala#L107